### ISSUE TITLE
ERPNext upgrade makes assets fail

### ISSUE THREAD

**Issue by sorinpopescu** [frappe/frappe_docker#1353]
## Steps to reproduce the issue

1. Install ERPNext with version v15.16.2
2. Finish instalation and go the application
3. Change version in .env (or docker compose) from v15.16.2 to v15.17.0
4. Rebuild container
5. Refresh works - because of cache - try incognito browser
6. Application works, but all assets are broken (css / js) 
7. Running migrate or build not helping
8. Fresh install works with database export / import but it's not the desired upgrade

### Stacktrace / full error message if available

```
![image](https://github.com/frappe/frappe_docker/assets/2951110/d70c465d-bc6b-4977-be13-788532cc1216)
```

**Comment by aynugek** [comment]
I have the same issue, however for me it's that either the browser loads old assets after running `bench build` or it can load `js` assets fine but `css` assets raise a `404`. I'm using frappe_docker to run `frappe v14.66.3` and `erpnext v14.64.0`.


![Screenshot_20240314_115953](https://github.com/frappe/frappe_docker/assets/59009890/9737d5ea-7332-4bca-8080-78d101a5ee86)

![Screenshot_20240314_115737](https://github.com/frappe/frappe_docker/assets/59009890/ca345164-b6c1-4605-a54f-0ec81cbbf512)

**Comment by aynugek** [comment]
Looks like removing the container and starting it as explained [here](https://github.com/frappe/erpnext/issues/38888#issuecomment-1897731812) fixes the issue.

**Comment by letajmal** [comment]
You should not run `bench build` command in docker production setup. Apps will be build during docker image build, there is no need and running bunch build will break your site. Deleting and redeploying the current containers will solve the issue(docker compose down && docker compose up -d)

**Comment by oleynikd** [comment]
@letajmal but rebuilding image with adding new or removing old apps will also break the site. It updates the `sites/` directory which is mapped to a shared volume. If we remove this volume all data will be lost, if we keep it - we'll get a 500 (Internal Server Error).
So how should we add/remove/update apps on a working production docker containers?

**Comment by letajmal** [comment]
Building a new image with the latest apps will not update anything under the sites folder. 

What's happening in the docker image is they're just initialising a bench and get-app the required apps.

Once the new image is built, you just need to migrate and clear cache for the changes to rollout into your deployment

**Comment by oleynikd** [comment]
@letajmal thank you. But what if I need to add new app? Or remove unused one?

**Comment by letajmal** [comment]
Remove that app from apps.json or from Dockerfile

**Comment by oleynikd** [comment]
@letajmal previously you said that

> Building a new image with the latest apps will not update anything under the sites folder.

But all assets are stored in `sites/assets/...` so after building new frappe+erpnext docker image the frontend will look for new asset files like `assets/frappe/dist/css/website.bundle.BHRNN4K5.css` and will not find them, because those build hashes changed (as well as file contents), but because the `sites` directory is mapped volume it is not updated when new container version starts up.
Please help to solve this issue.

**Comment by oleynikd** [comment]
So I had to start a new container (from newly built image) with shell, copy the contents of `sites/assets/assets.json` and `sites/assets/assets-rtl.json`, then go to my mapped `sites/assets` volume and update those files in mapped volume, then `bench clear-cache && bench clear-website-cache`, restarted my containers and now everything works fine.

**Comment by oleynikd** [comment]
By the way, the `apps.json` file will also never be updated because it is located in `sites` directory

**Comment by letajmal** [comment]
> So I had to start a new container (from newly built image) with shell, copy the contents of `sites/assets/assets.json` and `sites/assets/assets-rtl.json`, then go to my mapped `sites/assets` volume and update those files in mapped volume, then `bench clear-cache && bench clear-website-cache`, restarted my containers and now everything works fine.

Assets inside the sites directory is a sym link points to the apps directory

**Comment by oleynikd** [comment]
@letajmal but `sites/assets/assets.json` and `sites/apps.json` are not.

**Comment by letajmal** [comment]
> So I had to start a new container (from newly built image) with shell, copy the contents of `sites/assets/assets.json` and `sites/assets/assets-rtl.json`, then go to my mapped `sites/assets` volume and update those files in mapped volume, then `bench clear-cache && bench clear-website-cache`, restarted my containers and now everything works fine.

I am sorry if i mislead you, these files will also get updated. For an example, from the official [docs](https://frappeframework.com/docs/user/en/basics/sites#apps-txt)
>apps.txt 
This file contains a list of Frappe apps. An app should be listed here before it can be installed on a site. It is added automatically when bench get-app command or bench new-app command is run.

so apps.txt will get updated when the we bench get-app during a docker image build
[refer](https://github.com/frappe/frappe_docker/blob/main/images/production/Containerfile#L105) line 105

>RUN bench init \
  --frappe-branch=${FRAPPE_BRANCH} \
  --frappe-path=${FRAPPE_PATH} \
  --no-procfile \
  --no-backups \
  --skip-redis-config-generation \
  --verbose \
  /home/frappe/frappe-bench && \
  cd /home/frappe/frappe-bench && \
  bench get-app --branch=${ERPNEXT_BRANCH} --resolve-deps erpnext ${ERPNEXT_REPO} && \
  echo "{}" > sites/common_site_config.json && \
  find apps -mindepth 1 -path "*/.git" | xargs rm -fr
  
I assume all the other files you mentioned will also get updated during this step. May be someone with deep frappe framework knowledge can help us out here
  
Generally you should not worry about this, I have not came across any kinds of errors related to this

**Comment by letajmal** [comment]
![image](https://github.com/user-attachments/assets/c828fa2c-f38f-43b3-a1d5-ee251e220edc)



> So I had to start a new container (from newly built image) with shell, copy the contents of `sites/assets/assets.json` and `sites/assets/assets-rtl.json`, then go to my mapped `sites/assets` volume and update those files in mapped volume, then `bench clear-cache && bench clear-website-cache`, restarted my containers and now everything works fine.

these files are already present in frappe/erpnext image even before creating a new site

**Comment by oleynikd** [comment]
Yes, and they are moved to docker shared volume. So after building new docker image and running a new container those files will not be updated.

**Comment by ronyantony-ig** [comment]
I had the same issue. On my cicd I have 5 builds and the assest from the first one was being requsted by the browser, despite the container had the lastest in assets.json.

eg: container assets.json: "login.bundle.css": "/assets/frappe/dist/css/login.bundle.BNLXGCOL.css"
      broser was trying to fetch login.bundle.VNLPGCOL.css.

The possible issue with my environmnet was the  Redis cache holding old stale outdated assets.json data.

Running these commands solved the issue:

docker exec -it <redis-cache-container> redis-cli FLUSHALL
docker exec -it <redis-queue-container> redis-cli FLUSHALL

**Comment by legolas108** [comment]
> So I had to start a new container (from newly built image) with shell, copy the contents of `sites/assets/assets.json` and `sites/assets/assets-rtl.json`, then go to my mapped `sites/assets` volume and update those files in mapped volume, then `bench clear-cache && bench clear-website-cache`, restarted my containers and now everything works fine.

@oleynikd's procedure had worked for me also for quite a while, but recently again old `.css` files were requested in spite of following it, and this time what helped was clearing the Redis cache: `docker exec -it -u root redis-cache redis-cli FLUSHALL` (`redis-cache` being the name of the Redis cache container).

Just now see that @ronyantony-ig found this out already 2 hours ahead of me, but the commands seem to be missing the container name.

**Comment by malnajdi** [comment]
We are facing similar issue and we did run the redis flushall but same issue. This issue is rending the ui as not usable as all. 
any fix ?

**Comment by gparent** [comment]
> I had similar issue. On my cicd I havd 5 migaes from deployments and the assetsfrom the first one was being requsted by the browser, despite the running container had the latest in assets.json.
> 
> eg: container assets.json: "login.bundle.css": "/assets/frappe/dist/css/login.bundle.BNLXGCOL.css" broser was trying to fetch login.bundle.VNLPGCOL.css.
> 
> The possible issue with my environmnet was the Redis cache holding old stale outdated assets.json data.
> 
> Running these commands solved the issue:
> 
> docker exec -it redis-cli FLUSHALL docker exec -it redis-cli FLUSHALL

I do not know why this works, but it does. I'm using the `frappe_docker` repo and making a custom app.

Same symptoms as you had. javascript files were loading fine, css files from Frappe were not (getting a 404). I never ran bench building commands in my container or deleted the assets, and they were in the container anyway, just not being loaded because the app insisted on loading the wrong CSS file hash.

When I deploy an upgrade, I rebuild the docker containers, "down" the containers for my site, "up" them again, then run `bench migrate; bench clear-cache`. Not sure if this is the correct procedure but it has always been reliable up until this week. I do not restart or mess with Traefik.

After some frustrating troubleshooting, clearing the redis cache fixed it instantly.

Any idea what changed, if anything?

**Comment by ptrmtb** [comment]
> I had similar issue. On my cicd I havd 5 migaes from deployments and the assetsfrom the first one was being requsted by the browser, despite the running container had the latest in assets.json.
> 
> eg: container assets.json: "login.bundle.css": "/assets/frappe/dist/css/login.bundle.BNLXGCOL.css" broser was trying to fetch login.bundle.VNLPGCOL.css.
> 
> The possible issue with my environmnet was the Redis cache holding old stale outdated assets.json data.
> 
> Running these commands solved the issue:
> 
> docker exec -it redis-cli FLUSHALL docker exec -it redis-cli FLUSHALL

Thanks! This worked wonder for me

**Comment by systonium** [comment]
> > I had similar issue. On my cicd I havd 5 migaes from deployments and the assetsfrom the first one was being requsted by the browser, despite the running container had the latest in assets.json.
> > eg: container assets.json: "login.bundle.css": "/assets/frappe/dist/css/login.bundle.BNLXGCOL.css" broser was trying to fetch login.bundle.VNLPGCOL.css.
> > The possible issue with my environmnet was the Redis cache holding old stale outdated assets.json data.
> > Running these commands solved the issue:
> > docker exec -it redis-cli FLUSHALL docker exec -it redis-cli FLUSHALL
> 
> I do not know why this works, but it does. I'm using the `frappe_docker` repo and making a custom app.
> 
> Same symptoms as you had. javascript files were loading fine, css files from Frappe were not (getting a 404). I never ran bench building commands in my container or deleted the assets, and they were in the container anyway, just not being loaded because the app insisted on loading the wrong CSS file hash.
> 
> When I deploy an upgrade, I rebuild the docker containers, "down" the containers for my site, "up" them again, then run `bench migrate; bench clear-cache`. Not sure if this is the correct procedure but it has always been reliable up until this week. I do not restart or mess with Traefik.
> 
> After some frustrating troubleshooting, clearing the redis cache fixed it instantly.
> 
> Any idea what changed, if anything?

Did you find any answer for this? I'm facing a similar problem, where things are working fine for a week or two, then suddenly app tries to look for different hash files.

**Comment by gparent** [comment]
> > > I had similar issue. On my cicd I havd 5 migaes from deployments and the assetsfrom the first one was being requsted by the browser, despite the running container had the latest in assets.json.
> > > eg: container assets.json: "login.bundle.css": "/assets/frappe/dist/css/login.bundle.BNLXGCOL.css" broser was trying to fetch login.bundle.VNLPGCOL.css.
> > > The possible issue with my environmnet was the Redis cache holding old stale outdated assets.json data.
> > > Running these commands solved the issue:
> > > docker exec -it redis-cli FLUSHALL docker exec -it redis-cli FLUSHALL
> > 
> > 
> > I do not know why this works, but it does. I'm using the `frappe_docker` repo and making a custom app.
> > Same symptoms as you had. javascript files were loading fine, css files from Frappe were not (getting a 404). I never ran bench building commands in my container or deleted the assets, and they were in the container anyway, just not being loaded because the app insisted on loading the wrong CSS file hash.
> > When I deploy an upgrade, I rebuild the docker containers, "down" the containers for my site, "up" them again, then run `bench migrate; bench clear-cache`. Not sure if this is the correct procedure but it has always been reliable up until this week. I do not restart or mess with Traefik.
> > After some frustrating troubleshooting, clearing the redis cache fixed it instantly.
> > Any idea what changed, if anything?
> 
> Did you find any answer for this? I'm facing a similar problem, where things are working fine for a week or two, then suddenly app tries to look for different hash files.

Yeah it has been fixed now. If I remember correctly a redis volume that shouldn't have been used kept the old files. To fix it you can update your pull of this repo, then run the compose step again for the backend, then you can delete the unused redis volume.

**Comment by davispuh** [comment]
I'm still encountering this issue (latest v2.2.0 and v16.11.0) but I'm using my own Podman quadlet containers.
I'm building image with:
```
$ podman build --file images/custom/Containerfile --build-arg APPS_JSON_BASE64=... --build-arg FRAPPE_BRANCH=version-16 .
```

For the first time install I start it and inside container I run
```
$ bench new-site --install-app erpnext --set-default mysite
$ bench --site mysite install-app hrms
```

Then when updating I rebuild image and restart containers and do:
```
bench --site mysite migrate
```

But looks like this is not enough because `assets.json` is still referencing old names.
How exactly it's supposed to be updated?
Image has:
```json
"Volumes": {
    "/home/frappe/frappe-bench/logs": {},
    "/home/frappe/frappe-bench/sites": {},
    "/home/frappe/frappe-bench/sites/assets": {}
},
```
so they persist.

Now I tried:
```
$ bench clear-cache
$ bench clear-website-cache
$ valkey-cli FLUSHALL
```
And that resolved it but not sure which step exactly. Also note that I'm using Valkey from host, not inside container so it doesn't get restarted.

**Comment by TeddyViberg** [comment]
I had a lot of issues with this too and solved it by putting in a redis-cache redis-cli FLUSHDB in by build script. The fact that clear-cache and clear-website-cache doesn't seem to touch the redis-cache make it a bit tricky to debug because "I've already cleared the cache A THOUSAND TIMES" - redis is not top of mind for us vibe programmers hehe. 
When watching the docker build you can(I* can) usually see the warning "WARN Cannot connect to redis_cache to update assets_json", because redis is not running while building and therefore cannot be updated. I suspect this is the root cause of the error.

**Comment by davispuh** [comment]
I think touching Redis at build time is bad idea. Because when I build new image then Redis is used for currently running version.
I guess more sense would be after updating container to use new image then flash cache on Redis but `FLASHALL` / `FLUSHDB` is like nuclear option. There should be better way.

**Comment by davispuh** [comment]
I found the issue. `assets.json` file is being persisted in Redis key `assets_json`.
This means that after updating `assets.json` file is regenerated on disk, but because Redis key is not invalidated we still get served old file.
`bench clear-website-cache` should also clear it but currently it doesn't do that so it's a bug that needs fixing.
Something like this:
```diff
diff --git a/frappe/website/utils.py b/frappe/website/utils.py
--- frappe/website/utils.py
+++ frappe/website/utils.py
@@ -386,8 +386,9 @@
 		"website_full_index",
 		"languages_with_name",
 		"languages",
 		"website_404",
+		"assets_json",
 	]
 
 	if path:
 		frappe.cache.hdel("website_redirects", path)
```

Currently as a workaround after update need to do this:
```
$ redis-cli DEL assets_json
```

This fixes this issue.

Note that there is `clear_global_cache()` function in `frappe/cache_manager.py` that does clear this key, but it's not exposed as a command in `bench`.

**Comment by davispuh** [comment]
I wrote a fix https://github.com/davispuh/frappe/commit/350843d7e2347624c88ba4cf8399c67ab3aad735 but weirdly `frappe` repo doesn't let me create a PR...

**Comment by gparent** [comment]
> I wrote a fix [davispuh/frappe@350843d](https://github.com/davispuh/frappe/commit/350843d7e2347624c88ba4cf8399c67ab3aad735) but weirdly `frappe` repo doesn't let me create a PR...

https://github.com/frappe/frappe/issues/38145

TL;DR: You can submit your changes, just not as a PR out of the blue.

**Comment by DanielRadlAMR** [comment]
@davispuh I just ran into this issue as well. Is this really still not fixed? Have you proposed your change to frappe/frappe?

**Comment by davispuh** [comment]
> [@davispuh](https://github.com/davispuh) I just ran into this issue as well. Is this really still not fixed? Have you proposed your change to frappe/frappe?

Yeah it's still not fixed and my proposal is here in this issue, someone from Frappe needs to merge fix.

**Comment by DanielRadlAMR** [comment]
@davispuh Could you put your changes on a new branch and give me some insight into what this actually does? For example, when is this cache manager called? Is it triggered by `bench clear-cache`?

I'll try to get it merged.

**Comment by davispuh** [comment]
Fix is in [assets](https://github.com/davispuh/frappe/tree/assets) branch. 

>  when is this cache manager called? Is it triggered by bench clear-cache?

Looks like call stack is:
```
$ bench clear-cache
-> frappe/commands/utils.py#clear_cache()
  -> frappe.clear_cache() => frappe/cache_manager.py#clear_cache()
    -> 'else:  # everything' branch, which doesn't include clearing `assets_json`
  -> frappe/website/utils.py#clear_website_cache()
    -> clear_cache()
```

With my fix `assets_json` will be cleared in `utils.py#clear_cache()` part.

**Comment by DanielRadlAMR** [comment]
Hmm, looking at your changes, you actually deleted two lines that look like they would clear `assets_json`. I'll need to dig into it a bit further before proposing a change to Frappe.

**Comment by davispuh** [comment]
> Hmm, looking at your changes, you actually deleted two lines that look like they would clear `assets_json`. I'll need to dig into it a bit further before proposing a change to Frappe.

Yeah because that's wrong place where to do it. That path doesn't run on `bench clear-cache`. It only runs when clearing permission cache which doesn't make sense to clear assets because permissions has nothing to do with assets...

**Comment by DanielRadlAMR** [comment]
Created an Issue in frappe/frappe: https://github.com/frappe/frappe/issues/39812

**Comment by DanielRadlAMR** [comment]
@davispuh  https://github.com/frappe/frappe/pull/39986

Should be fixed with the next Release
